### ISSUE TITLE
[question] is conan impossible to use in a monorepo?

### ISSUE THREAD

**Issue by blackliner** [conan-io/conan#8995]
## Question

Is it possible to use conan in a monorepo to manage multiple packages?

## Current situation

We have a monorepo with one huge CMake based superproject and use conan as a consumer only. We want to split this superproject up into separate conan packages (each containing multiple libraries/CMake targets), that have a certain dependencies to each other. Those packages shall have their conan recipe right with the code ([aka in-source recipe](https://docs.conan.io/en/latest/creating_packages/package_repo.html)), because it makes sense, right? Conan will be the **native build** infrastructure for those packages, so there will not be any other `build.sh` **per package**, maybe one for the whole monorepo.

The "products" of this monorepo will be executables and libraries that go to customers, and conan packages which will be consumed by other departments in the same company. This might be relevant to udnerstand the need for versioning of packages.

### RREV

With in-source recipes, **any** change in the recipe **or** the sourcecode will result in a new RREV. This will propagate throughout the whole dependency tree, which sounds like a lot of maintainance/tooling required. Assume the following dependency graph:
![image](https://docs.conan.io/en/latest/_images/conan-info_deps_html_graph.png)

Lets assume all those packages represent in-source recipes. When a new PR is created to change even a comment in `zlib`, a new RREV is calculated for this package. `poco` now needs to be updated, because it had the old RREV as part of the `self.requires` call. The same needs to happen for `pcre`, and since its RREV also changed, we need to go back to `poco`'s recipe and update the RREV to `pcre` too. This will become maintainence hell 😞
One idea I had was to come up with some advanced tooling (like a python package or a `python_requires` conan package), that will abstract the dependencies away by calling something like `tool.requires(pcre)` (no version, no RREV) and a **centrally** defined list of package/version/RREV combination, for example at the repo root. But this now breaks the conan build process, because this information is not available when the sourcecode is copied to the conan local build cache

### Versioning

The common monorepo approach does not see it required/possible to have independend versioning of you libraries, as the whole repo must have a consistent state. Versions are defined in source (could be the conanfile, or reading from a version.txt). But conan very much relies on version numbers for caching, so I see some conflicting solutions:

a) each package has version 1.0, and the depend to each other with just `self.requires("package_a/1.0.0")`: This will probably disable conans caching feature, and create packages with the same version, but different RREV's. This may not be an issue if you always export all your recipes before starting to build anything, but you very likely will get problems when an conan repo comes into play and provides hundreds of alternative packages with the same version number. So lets use RREV? 😞 see above, and also we would need a solution to provide "real" versioning to packages that represent our external interface (aka "products")

b) each package will have its own, independend version. This means that for **any** change you must update the version you changed, and consequently all packages that depend on this changed package, because if their dependency changes, their `self.requires` changes, which is a change in the source. This is the same issue as with RREV's above.

---
- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by demorgan104** [comment]
Hello @blackliner , I was looking at this issue and was thinking, your original problem is that you cannot build zlib for example without the header files of pcre, right ?  And other header files located in your super-repo ?

In this case, have you though about creating packages that contain only the header files of your super-repo ?

In this way, you will deliver libraries and re-compilation of these libraries is happening only when certain header files are changed, and only the libraries depending on these header files will be changed :)

Of course, in this way, you could step by step try to re-factor the source code.

**Comment by blackliner** [comment]
hey @demorgan104, I think you misunderstood, this is a conceptual problem. Building `zlib` or `pcre` was never the issue, it is about how should versions and RREV's be handled in a monorepo, multi-package project.

Our current solution is:
- every package has version `1.0.0` and that wont change
- every package refers to any other internal package (one within the monorepo) by `name/1.0.0`
- before a built is started, a script exports all package recipes to the local cache. This way, conan is forced to use the current state of the repo as-is, and since binaries are chached through RREV and package ID, caching works just fine.
- external packages, like OpenCV, are referenced with `name/version#RREV`

**Comment by demorgan104** [comment]
Hello @blackliner,

Sorry for my misunderstanding. 

I have some extensive experience with Conan for embedded systems where such features are required to ensure reproducibility and to be able to reproduce a binary at any time from its original source code.

I also know the pain I got once when we messed up with Conan's cache:

- we got a perfect build
- the software was running on the HW with sporadic issues (a communication protocol was simply not performing the handshake)
- the solution was to clean the cache and rebuild everything, suddenly everything worked again on the HW

Unfortunately, it is a bit hard for me to understand your case because mono-repo and packages are mixed.

Also because I guess you are talking about the mono-repo as a local internal development solution.

For example for this statement: "We want to split this superproject up into separate conan packages (each containing multiple libraries/CMake targets), that have a certain dependencies to each other" the certain dependencies should never be cycle dependencies. If you are really ready to split the source code in packages, than you can do it and create conan packages out of it. You can start splitting in 2 first, then the resulting repos in another two parts and so on....
To develop you can use workspace features....if you use CMake find_packages feature than you can instruct CMake to find your packages after you assemble the mono-repo again.

I can help if you provide clarifications to the following:

- what do you mean by Conan inside mono-repo ? Conan is designed exactly to prevent that (and you have this feature to help for local development: https://docs.conan.io/en/latest/developing_packages/workspaces.html)
- How is your git/versioning system structured ?
- Please make a clear delimitation between local development and delivery pipelines :)


Of course a git repo as an example will be more than welcomed :)

**Comment by pviolette3** [comment]
He @blackliner - quick question - did you ever figure this conan-in-monorepo challenge out, or did you move to something else besides conan?

**Comment by opajonk** [comment]
I would also be interested in this - since we're also looking at "mono-repo + conan 2", and it seems to be not totally straightforward.

**Comment by thorsten-klein** [comment]
At my work we have solved this by introducing a new method called "self.requires_local('../mylib')".
Within this method we export the recipe into the local cache, then we read the resulting reference and call "self.requires(reference)". We also append this reference to conandata.yml on the fly (required if the recipe is itself is exoorted). This works perfectly for us, even if conan maintainers warn about running "conan export" during a conan install.

**Comment by memsharded** [comment]
The current recommendation to work with mono-repo like projects would be:

- First, consider not using Conan to split the mono repo. Conan is a package manager, and it is designed to create and manage packages, which are basically independent units of development and release, each one with its own independent version. A mono-repo is exactly the opposite of component/package based development paradigm. Conan can be used for the third party external dependencies to the mono repo as usual.
- If still want to do both development paradigms at once (mono-repo + packages), then the recommendation would be to use ``editable`` packages. Put every folder/subproject in the mono-repo as ``conan editable add``, then pick the most downstream consumer and ``conan install --build=editable`` it for orchestrating the build
- You might want to automate the management of the monorepo with some script that manages the ``editables`` more automatically, specially if there are many. This is basically the "worskspace" concept. We have already started to work on it to bring it to Conan 2, but it might take a while, in the meantime a custom script or a Conan custom command might help.

**Comment by opajonk** [comment]
As usual - thanks for the very quick answer :-)

> First, consider not using Conan to split the mono repo. Conan is a package manager, and it is designed to create and manage packages, which are basically independent units of development and release, each one with its own independent version. A mono-repo is exactly the opposite of component/package based development paradigm. Conan can be used for the third party external dependencies to the mono repo as usual.

Of course, you are somewhat right - mono-repo and package management in one looks "strange". However, I see this valid use-case: if the modules (packages) handled in the mono-repo are mutually dependent, developed & released together, but upstream customers are free to use some or all of them - with the constraint that they have been released (and validated!) together.

**Comment by memsharded** [comment]
Conan 2 has summarized the "workspace" feature. See current "incubating feature" in https://docs.conan.io/2/incubating.html

The ticket centralizing efforts is https://github.com/conan-io/conan/issues/15992

I am closing this ticket, please create new tickets with details regarding the new workspace feature, thanks for the feedback.
